dashboard/rendering.py

```diff
@@ -14,40 +26,153 @@ def render_login_page() -> tuple:
+
+        highest_price = get_price_info(filtered_df, False)
+        st.sidebar.write(
+            f'Highest Price: £{round(highest_price["Price"], 2)}')
+
+        lowest_price = get_price_info(filtered_df, True)
+        st.sidebar.write(f'Lowest Price: £{round(lowest_price["Price"], 2)}')
+
+        # Changes the Product Availability from True/False to In Stock/Out of Stock.
+        if most_recent_price["Product Availability"] is True:
+            st.sidebar.write(
+                'Availability: In Stock')
+        else:
+            st.sidebar.write(
+                'Availability: Out of Stock')
+
+        # Button that links to the products URL.
+        if st.sidebar.button("Go to Product"):
+            webbrowser.open_new_tab(most_recent_price["Product URL"])
+
     else:
         st.sidebar.write("No image available for the selected product.")
 
 
 def display_admin_main_body(df: DataFrame) -> None:
-    head_cols = st.columns(3)
+    """
+    Displays all of the admin main body for streamlit.
+    """
+    sorted_df = df.sort_values(
+        by=['Product ID', 'Updated At'], ascending=[True, False])
+
+    most_recent_prices = sorted_df.groupby('Product ID').first()
 
+    # Header metrics
+    head_cols = st.columns(3)
     with head_cols[0]:
         st.metric("Total No. of Users :bust_in_silhouette:",
                   df["User Email"].nunique())
+
     with head_cols[1]:
         st.metric("Total No. of Products", df["Product Name"].nunique())
 
     with head_cols[2]:
         st.metric("Total No. of Subscriptions",
                   df["Subscription ID"].nunique())
 
+    # Main body of Dashboard
+    body_cols = st.columns(2)
+    with body_cols[0]:
+
+        name_in_selected_products_price = get_names_of_selected_products(
+            most_recent_prices, "most_recent")
+
+        if not name_in_selected_products_price.any():
+            st.error("Please select at least one product.")
+
+        else:
+            st.altair_chart(get_latest_price_readings(
+                most_recent_prices[name_in_selected_products_price]), use_container_width=True)
+
+    with body_cols[1]:
+        # Need to be repeated due to the use of a different dataframe.
+        name_in_selected_products = get_names_of_selected_products(
+            df, "all_admin_pop")
+
+        if not name_in_selected_products.any():
+            st.error("Please select at least one product.")
+        else:
+            st.altair_chart(get_popularity_of_products(df[name_in_selected_products]),
+                            use_container_width=True)
+
+    # Repeated in order for the selection bar to look more presentable.
+    name_in_selected_products_all = get_names_of_selected_products(
+        df, "all_admin")
+
+    st.altair_chart(get_price_of_products_over_time(df[name_in_selected_products_all]),
+                    use_container_width=True)
+
+
+def display_user_admin_info(users: list[dict]) -> None:
```

Comment:
Great job having functions to separate the admin and user page - makes it all so much more readable